profiler/src/ProfilerEngine/Datadog.Profiler.Native/NetworkProvider.cpp

```diff
@@ -0,0 +1,522 @@
+    if ((result->GetFramesCount() == 0) && (failureCount % 100 == 0))
+    {
+        // log every 100 failures
+        failureCount++;
+        Log::Warn("Failed to walk ", failureCount, " stacks for sampled HTTP request: ", HResultConverter::ToStringWithCode(hrCollectStack));
+        return false;
+    }
+
+    info.AppDomainId = threadInfo->GetAppDomainId();
+    info.LocalRootSpanID = result->GetLocalRootSpanId();
+    info.SpanID = result->GetSpanId();
+    info.StartCallStack = result->GetCallstack();
+    info.StartThreadInfo = threadInfo;
+
+    return true;
+}
+
+void NetworkProvider::OnRequestStart(std::chrono::nanoseconds timestamp, LPCGUID pActivityId, std::string url)
+{
+    NetworkActivity activity;
+    if (!TryGetActivity(pActivityId, activity))
+    {
+        return;
+    }
+
+    auto existingInfo = _requests.find(activity);
+    if (existingInfo != _requests.end())
+    {
+        // TODO: this should never happen; i.e. a request with the same activity is already in progress
+        _requests.erase(activity);
+
+        // we start a new request with the same activity to avoid losing both the previous and the current one
+    }
+
+    auto slot = _requests.insert_or_assign(activity, NetworkRequestInfo{ url, timestamp });
+    if (!CaptureThreadInfo(slot.first->second))
+    {
+        _requests.erase(activity);
+    }
+}
+
+void NetworkProvider::OnRequestStop(std::chrono::nanoseconds timestamp, LPCGUID pActivityId, uint32_t statusCode)
+{
+    NetworkActivity activity;
+    if (!TryGetActivity(pActivityId, activity))
+    {
+        return;
+    }
+
+    auto requestInfo = _requests.find(activity);
+    if (requestInfo == _requests.end())
+    {
+        // skipped request
+        return;
+    }
+
+    // sampling can be forced for tests
+    if (!_pConfiguration->ForceHttpSampling())
+    {
+        // only requests lasting more than a threshold are captured
+        if ((timestamp - requestInfo->second.StartTimestamp).count() > _requestDurationThreshold)
+        {
+            // skip this request
+            _requests.erase(activity);
+            return;
+        }
+
+        // TODO: add additional filtering to avoid too many samples
+        //       e.g. requests with specific status codes, min duration per phase, dynamic such as for exceptions, etc.
+    }
+
+    RawNetworkSample rawSample;
+    FillRawSample(rawSample, requestInfo->second, timestamp);
+    rawSample.StatusCode = statusCode;
+    if (!requestInfo->second.Error.empty())
+    {
+        rawSample.Error = std::move(requestInfo->second.Error);
+    }
+
+    if (!requestInfo->second.HandshakeError.empty())
+    {
+        rawSample.HandshakeError = std::move(requestInfo->second.HandshakeError);
+    }
+
+    if (requestInfo->second.Redirect != nullptr)
+    {
+        // will be an empty string for .NET 7
+        rawSample.RedirectUrl = std::move(requestInfo->second.Redirect->Url);
+        rawSample.HasBeenRedirected = true;
```

Comment:
do we need this flag ?
if `rawSample.RedirectUrl` is not empty it means it has been redirected no?
Is there a case where `rawSample.RedirectUrl` is empty but a redirection occured?

Reply:
Yes: the redirect url will be empty in .NET 7 even if a redirect (that I'm detecting) happens